python/cugraph/cugraph/dask/sampling/uniform_neighbor_sample.py

```diff
@@ -184,6 +189,13 @@ def uniform_neighbor_sample(
             df['hop_id']: dask_cudf.Series
                 Contains the hop ids from the sampling result
     """
+    if _multiple_clients:
+        # Distributed centralized lock to allow
+        # two disconnected processes (clients) to coordinate a lock
+        # https://docs.dask.org/en/stable/futures.html?highlight=lock#distributed.Lock
+        lock = Lock("sampling_mg_lock")
+        lock.acquire(timeout=1)
```

Comment:
The lock isn't actually needed until the `client.submit()` call that runs sampling below, is that correct?  I'm just wondering if the acquire should be moved closer to that call to maximize parallelism by holding the lock as little as possible.

Reply:
Again, I am not sure how safe our renumbering is . I can add it and release it across a bunch of places if that helps.

Reply:
I share Rick's sentiment here. As someone who isn't super-familiar with the sampling code, it's unclear to me exactly which resources are being protected from the lock and which operations depend on those resources.

Also, what about possible exceptions between `lock.acquire` and `lock.release`? I think it would be best practice to use it as a context manager like `with Lock("sampling_mg_lock"): ...`

Reply:
Alright, so if I understand correctly, this lock is used to ensure that a dask cluster will only perform sampling for one dataset at a time, right? Even if two clients want to sample from two different graphs, they won't be able to sample at the same time. To reiterate my above thoughts, I think _all_ locks should indicate what resources they are protecting, even if it's as broad as "cluster usage".

Reply:
> Alright, so if I understand correctly, this lock is used to ensure that a dask cluster will only perform sampling for one dataset at a time, right? Even if two clients want to sample from two different graphs, they won't be able to sample at the same time. 

Yes, my current guess of the problem is that our PLC layer is not multi-client safe so if we submit tasks to that layer we seem to hang.   Tried to address those with the latest push . 

Thanks for the feeback @eriknw  and @rlratzel . All great points that i agree with. Thanks for catching those subtleties. Please take a look and see if the last push solves all the raised points.   Just for completeness, i tested it with new changes and it seems to be working.

Reply:
LGTM! 👍